src/nwp_emulator/grib_file_reader.cc

```diff
@@ -0,0 +1,302 @@
+    currentFile_ = nullptr;
+    grib_ = nullptr;
+}
+
+bool GRIBFileReader::hasReadAll() {
+    return (step_ >= srcFilenames_.size() || step_ >= stepCountLimit_);
+}
+
+bool GRIBFileReader::openNextFile() {
+    if (rank_ != root_) {
+        eckit::Log::warning() << "Only root reader is allowed to read files, skipping..." << std::endl;
+        return false;
+    }
+    bool result = true;
+    // Open the GRIB file corresponding to the current step
+    if (!openGribFile(srcFilenames_[step_])) {
+        // Skip this file
+        eckit::Log::error() << "An error occurred while reading: " << srcFilenames_[step_] << std::endl;
+        closeGribFile();
+        return false;
+    }
+    // Run some validity checks on the file
+    char buffer[64];
+    size_t size = sizeof(buffer);
+    codes_get_string(grib_, "gridName", buffer, &size);
+    if (std::string(buffer) != gridName_) {
+        eckit::Log::error() << "Data for step " << step_ << "are on grid " << buffer << " instead of " << gridName_
+                            << ", skipping step..." << std::endl;
+        result = false;
+    }
+    if (!result) {
+        // Note: extracting this here in case more checks are run in the future
+        closeGribFile();
+        return result;
+    }
+}
+
+bool GRIBFileReader::readNextStep(std::vector<atlas::Field>& fields, std::vector<atlas::Field>& glb_fields) {
+    int result = 1;
+    // Open the current file or skip
+    if (rank_ == root_) {
+        result = static_cast<int>(openNextFile());
+    }
+    eckit::mpi::comm().broadcast(result, root_); // use int as bool can't be broadcast by eckit mpi
+    if (!result) {
+        ++step_;
+        return false;
+    }
+    // Broadcast the number of messages in the current file for looping
+    eckit::mpi::comm().broadcast(count_, root_);
+    char buffer[10];
+    size_t size = sizeof(buffer);
+    std::vector<char> vecBuffer(size);
+    for (int msg_id = 0; msg_id < count_; ++msg_id) {
+        std::memset(buffer, 0, 10);
+        size = sizeof(buffer);
+        vecBuffer.resize(size, '\0');
+        if (rank_ == root_) {
+            codes_get_string(grib_, "shortName", buffer, &size);
+            vecBuffer.assign(buffer, buffer + size);
+        }
+        eckit::mpi::comm().broadcast(vecBuffer, root_);
+        std::string param(vecBuffer.begin(), vecBuffer.end());
+        param.resize(strlen(param.c_str()));
+        // Find the field name corresponding to the current message and populate it with its data
+        auto it = std::find(params_.begin(), params_.end(), param);
+        if (it != params_.end()) {
+            std::variant<std::vector<float>, std::vector<double>> values; // support single and double precision
+            if (rank_ == root_) {
+                size_t nbOfValues;
+                codes_get_size(grib_, "values", &nbOfValues);
+                if (std::is_same<FIELD_TYPE_REAL, double>::value) {
```

Comment:
```suggestion
                if constexpr (std::is_same_v<FIELD_TYPE_REAL, double>) {
```
now values does not need to be a variant and can just be
```
std::vector<FIELD_TYPE_REAL> values;
```

Reply:
I tried your suggestion but compilation fails the same way it did before I used the variant:
```
std::vector<FIELD_TYPE_REAL> values(nbOfValues);
if constexpr (std::is_same_v<FIELD_TYPE_REAL, double>) {
    codes_get_double_array(grib_, "values", values.data(), &nbOfValues);
}
else {
    codes_get_float_array(grib_, "values", values.data(), &nbOfValues);
}
```
Throws the following error at compilation:
```
error: argument of type "double *" is incompatible with parameter of type "float *"
              codes_get_float_array(grib_, "values", values.data(), &nbOfValues);
```
Am I missing something ?

Reply:
Hi @cducher I took liberty to try to reproduce, and committed a working fix based on reinterpret_cast. It also avoids the copy into the temporary variant.

Reply:
Thanks Willem I'll give it a try !